### PR TITLE
feat: Store delete method

### LINKED ISSUE #45: store.Delete()
Would be good to have a `Delete()` method on the store interface.

Use case: transient storage, where a state machine needs to delete state to ensure a particular instance of the machine proceed properly on towards the next state.

### PR DESCRIPTION
**Pull request by DRK3**
closes #45 

Signed-off-by: Derek Trider <derek.trider@securekey.com>

### CODE REVIEW

File: pkg/storage/memstore/memstore.go [hunk 1]
```diff
@@ -129,3 +129,12 @@ func (m *MemStore) close() {
 
 	m.db = make(map[string][]byte)
 }
+
+// Delete deletes the key-value pair associated with k.
+func (m *MemStore) Delete(k string) error {
```
Comment (llorllale) [inline pkg/storage/memstore/memstore.go:134]: should return an error if `k` is not found

File: pkg/storage/mysql/mysqlstore.go [hunk 2]
```diff
@@ -333,6 +334,22 @@ func (s *sqlDBStore) Query(findQuery string) (storage.ResultsIterator, error) {
...
 }
 
+func (s *sqlDBStore) Delete(k string) error {
+	if k == "" {
+		return storage.ErrKeyRequired
+	}
+
+	//nolint: gosec
+	// TODO address SQL injection warning #38
+	_, err := s.db.Exec("DELETE FROM `"+s.tableName+"` WHERE `key`= ?", k)
```
Comment (llorllale) [inline pkg/storage/mysql/mysqlstore.go:344]: the result contains the number of rows affected. Should throw error if this number == 0

File: pkg/storage/couchdb/couchdbstore.go [hunk 3]
```diff
@@ -355,6 +359,42 @@ func (c *CouchDBStore) getStoredValueFromRawDoc(rawDoc map[string]interface{}, k
 	return strippedJSON, nil
 }
 
+func (c *CouchDBStore) getRevision(k string) (string, error) {
+	rawDoc, err := c.getRawDoc(k)
+	if err != nil {
+		return "", err
```
Comment (llorllale) [inline pkg/storage/couchdb/couchdbstore.go:365]: There are many unwrapped errors in this file

File: pkg/storage/memstore/memstore.go (hunk 1 above)
Comment (DRK3) [reply pkg/storage/memstore/memstore.go:134]: Done

File: pkg/storage/mysql/mysqlstore.go (hunk 2 above)
Comment (DRK3) [reply pkg/storage/mysql/mysqlstore.go]: Done

File: pkg/storage/couchdb/couchdbstore.go (hunk 3 above)
Comment (DRK3) [reply pkg/storage/couchdb/couchdbstore.go]: Done - for all the changed code, I've made sure the errors are wrapped now. For unwrapped errors in the rest of the file: I will push a separate PR to fix this in order to keep this PR more focused.

File: pkg/storage/store.go [hunk 4]
```diff
@@ -23,8 +23,11 @@ var ErrIndexingNotSupported = errors.New("indexing is not supported")
 // ErrQueryingNotSupported  is used when querying is not supported by the store implementation.
 var ErrQueryingNotSupported = errors.New("querying is not supported")
 
-// EndKeySuffix end key suffix
-const EndKeySuffix = "!!"
```
Comment (DRK3) [inline pkg/storage/store.go:27]: Note that `EndKeySuffix` was an unused const

File: pkg/storage/store.go [hunk 5]
```diff
@@ -68,6 +74,9 @@ type Store interface {
 	// Query queries the store for data based on the provided query string, the format of
 	// which will be dependent on what the underlying store requires.
 	Query(query string) (ResultsIterator, error)
+
+	// Delete deletes the key-value pair associated with k.
+	Delete(k string) error
```
Comment (rolsonquadras) [inline pkg/storage/store.go:79]: I'm wondering if `Remove` would make more sense as other APIs have Put/Get rather than Insert/Select.

Comment (DRK3) [reply pkg/storage/store.go:79]: Sounds reasonable to me...done!

Comment (llorllale) [reply pkg/storage/store.go]: This change makes no sense and I wouldn't be surprised if it is reverted in the future:

* redis: https://redis.io/commands/del
* memcache: https://www.tutorialspoint.com/memcached/memcached_delete_key.htm
* aries-framework-go's `Store` API
* CouchDB: https://docs.couchdb.org/en/stable/api/document/common.html#delete--db-docid

Comment (DRK3) [reply pkg/storage/store.go:79]: MySQL also uses `DELETE` (as can be seen in the code added here)... so it looks to me like the original "Delete" terminology was the best choice.

(it looks like some APIs may combine Put/Get with "Remove", but after a closer look I think this isn't universal...)

Reverted.

**Pull request merged**